File: gtci/03-sliding-window/cpp/7-permutation-in-a-string.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    bool findPermutation(const string &str, const string &pattern) {
        std::unordered_map<char, int> pattern_chars =
            Solution::constructFrequencyMapFromString(pattern);

        int windowStart = 0;
        std::unordered_map<char, int> chars;
        for (int windowEnd = 0; windowEnd < str.size(); windowEnd++) {
            int windowSize = windowEnd - windowStart + 1;

            chars[str[windowEnd]] += 1;

            while (windowSize > pattern.size()) {
                chars[str[windowStart]] -= 1;
                if (chars[str[windowStart]] == 0) chars.erase(str[windowStart]);

                windowStart++;
                windowSize = windowEnd - windowStart + 1;
            }

            if (chars == pattern_chars) return true;
        }

        return false;
    }

   private:
    static std::unordered_map<char, int> constructFrequencyMapFromString(
        const std::string &str) {
        std::unordered_map<char, int> chars;
        for (char ch : str) chars[ch] += 1;

        return chars;
    }
};

int main() {
    Solution sol;

    assert(sol.findPermutation("oidbcaf", "abc") == true);
    assert(sol.findPermutation("odicf", "dc") == false);
    assert(sol.findPermutation("bcdxabcdy", "bcdyabcdx") == true);
    assert(sol.findPermutation("aaacb", "abc") == true);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/03-sliding-window/python3/7-permutation-in-a-string.py
class Solution:
    def findPermutation(self, str1, pattern):
        pattern_chars = self.construct_freq_map_from_string(pattern)

        chars = {}
        max_window_size = len(pattern)

        window_start = 0
        for window_end in range(len(str1)):
            window_size = window_end - window_start + 1

            if str1[window_end] not in chars:
                chars[str1[window_end]] = 0
            chars[str1[window_end]] += 1

            if window_size > max_window_size:
                chars[str1[window_start]] -= 1
                if chars[str1[window_start]] == 0:
                    del chars[str1[window_start]]

                window_start += 1
                window_size = window_end - window_start + 1

            if chars == pattern_chars:
                return True

        return False

    def construct_freq_map_from_string(self, s):
        chars = {}

        for char in s:
            if char not in chars:
                chars[char] = 0
            chars[char] += 1

        return chars


if __name__ == "__main__":
    sol = Solution()

    assert sol.findPermutation("oidbcaf", "abc") == True
    assert sol.findPermutation("odicf", "dc") == False
    assert sol.findPermutation("bcdxabcdy", "bcdyabcdx") == True
    assert sol.findPermutation("aaacb", "abc") == True

    print("All test cases passed.")
